### DISCUSSION TITLE
What's the way to find the Window that references UserControl?

### DISCUSSION THREAD

**Discussion by IronEaterBeast** [AvaloniaUI/Avalonia#15765] [Q&A] [answered]
There is a button in UserControl, and I need to use ShowDialog to open a window when the button is clicked, but I haven't found a good way.

```
private void XApiSetting_Click(object? sender, RoutedEventArgs e)
{
    var apiWindow = new ApiSettings();
    //apiWindow.ShowDialog((MainWindow)(((Button)sender).GetVisualRoot()));
    //apiWindow.Show();
}
```

**Comment by thevortexcloud** [comment, marked as answer]
You can use the TopLevel APIs. 

https://docs.avaloniaui.net/docs/concepts/toplevel

EG

```csharp
var window = TopLevel.GetTopLevel(this) as Window;
```

Windows themselves inherit from TopLevel

https://github.com/AvaloniaUI/Avalonia/blob/42aa77ebdcdb39195e452fe564991eb9eed46d44/src/Avalonia.Controls/WindowBase.cs#L18

But you should probably still check if the top level is actually a window as there are a few things that implement it which you may run into depending on how your app runs. That said I have never seen it return anything but a Window when running a desktop app. Not sure what it returns on non desktop apps.

**Reply by IronEaterBeast** [reply]
Thank u

**Reply by workgroupengineering** [reply]
Please mark the @thevortexcloud's response as answer for quick identification.
